### PR TITLE
Introduce a dedicated parser for prometheus

### PR DESCRIPTION
**Pull request by DmitrySenin**
closes: #4414

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### Testing
Tested locally with the following configuration of input:
```
[[inputs.http_listener_v2]]
  service_address = ":8094"
  path = "/metrics/job/fqt"
  data_format = "prometheus"
```

### CODE REVIEW

**Comment by danielnelson** [comment]
Since `metric_version = 1` is deprecated now, I'd like to keep its use limited to backwards compatibility.  Do you think you could split the code so that v1 is only available on the prometheus input and the new parser only does v2?

**Comment by DmitrySenin** [comment]
@danielnelson , sure. Done

**Comment by DmitrySenin** [comment]
@danielnelson , Any chance to include it into 1.15? My team is looking forward to this functionality

**Comment by danielnelson** [comment]
Sorry, we have missed the merge window for 1.15, it will have to go in 1.16.

File: plugins/parsers/prometheus/parser_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+# TYPE get_token_fail_count counter
+get_token_fail_count 0
+`
+
+	validUniqueLine = `# HELP get_token_fail_count Counter of failed Token() requests to the alternate token source
+`
+
+	validUniqueSummary = `# HELP http_request_duration_microseconds The HTTP request latencies in microseconds.
+# TYPE http_request_duration_microseconds summary
+http_request_duration_microseconds{handler="prometheus",quantile="0.5"} 552048.506
```
Comment (platan) [inline plugins/parsers/prometheus/parser_test.go:26]: Should we add some tests for parsing metrics with timestamps?

**Comment by jakemcc** [comment]
Just stumbled across this and thought I'd add a comment that we'd also find this useful. Thanks for doing the work.

File: plugins/parsers/prometheus/parser_test.go (hunk 1 above)
Comment (DmitrySenin) [reply plugins/parsers/prometheus/parser_test.go:26]: If you have some scenarios, please, share and I can add them to the suit

**Comment by jakemcc** [comment]
Just checking in on this. Anything missing that I might be able to help with?

File: plugins/parsers/prometheus/parser_test.go (hunk 1 above)
Comment (jakemcc) [reply plugins/parsers/prometheus/parser_test.go:26]: @DmitrySenin I happened to be looking at the prometheus docs and realized there are examples of this ask there. [Here](https://prometheus.io/docs/instrumenting/exposition_formats/#text-format-example) you go.

Comment (DmitrySenin) [reply plugins/parsers/prometheus/parser_test.go:26]: Oh, got you. I was thinking that the ask was about "timestamp" values in labels. Sure, I will add new cases then

**Comment by jakemcc** [comment]
Not saying this to put any pressure on this being merged but just wanted to report I've been running a build of telegraf with this change and am successfully reporting metrics using it. Thanks for doing the work. 👏 👏 👏 

Thanks everyone for putting work in to both manage this project and submit PRs.

File: plugins/parsers/prometheus/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Prometheus Text-Based Format
+
+There are no additional configuration options for [Prometheus Text-Based Format][]. The metrics are parsed directly into Telegraf metrics. It is used internally in [prometheus input](/plugins/inputs/prometheus) or can be used in [http_listener_v2](/plugins/inputs/http_listener_v2) to simulate Pushgateway
```
[the remark this answers is not in the document]
Comment (DmitrySenin) [reply plugins/parsers/prometheus/README.md]: Should be fixed now

File: docs/DATA_FORMATS_INPUT.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ Protocol or in JSON format.
 - [Nagios](/plugins/parsers/nagios)
 - [Value](/plugins/parsers/value), ie: 45 or "booyah"
 - [Wavefront](/plugins/parsers/wavefront)
+- [Prometheus](/plugins/parsers/prometheus)
```
[the remark this answers is not in the document]
Comment (DmitrySenin) [reply docs/DATA_FORMATS_INPUT.md]: Should be good now

File: plugins/parsers/prometheus/parser_test.go (hunk 1 above)
Comment (DmitrySenin) [reply plugins/parsers/prometheus/parser_test.go:29]: Ops, I didn't read the docs so thoroughly. Thanks for point this out. Added a test for a counter with timestamp

**Comment by p-zak** [comment]
@DmitrySenin it seems that there is conflict with `plugins/inputs/prometheus/parser.go`

**Comment by BlueskyFR** [comment]
This should get merged!

**Review by ssoroka — APPROVED** [review]

**Review by reimda — APPROVED** [review]

**Pull request merged**